explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,252 @@
+(perf-pgo)=
+
+# Profile-Guided Optimization
+
+[Profile-guided optimization](https://en.wikipedia.org/wiki/Profile-guided_optimization) (PGO) or feedback-driven optimization (FDO) are synonyms for the same concept.  The idea is that a binary is compiled with built-in instrumentation which, when executed, will generate a profile.  This profile can be used as input for the compiler in a subsequent build of the same binary, and it will serve as a guide for further optimizations.
+
+It can be hard to do profiling of real world applications. Ideally, the profile should be generated by a representative workload of the program, but it's not always possible to simulate a representative workload.  Moreover, the built-in instrumentation impacts the overall performance of the binary which introduces a performance penalty.
+
+In order to address these problems, nowadays we use tools like `perf` to "observe" what the binary is doing externally (sampling it, by monitoring events using Linux kernel's PMU --- Performance Monitoring Unit), which makes the process more suitable to be used in production environments.  This technique works better than the regular built-in instrumentation, but it still has a few drawbacks that we will expand later.
+
+## Caveats
+
+* The purpose of this guide is to provide some basic information about what PGO is.  In order to do that, we will look at a simple example using OpenSSL (more specifically, the `openssl speed` command) and learn how to do basic profiling.  We will not go into a deep dive on how to build the project, and it is assumed that the reader is comfortable with compilation, compiler flags and using the command line.
+
+* Please note that, despite being a relatively popular technique, PGO is not always the best approach to optimize a program.  The profiling data generated by the workload will be extremely tied to it, which means that the optimized program might actually have worse performance when other types of workloads are executed.  There is not a one-size-fits-all solution for this problem, and sometimes the best approach might be to **not** use PGO after all.
+
+* If you plan to follow along, we recommend setting up a test environment for this experiment.  The ideal setup involves using a bare metal machine because it's the more direct way to collect the performance metrics.  If you would like to use a virtual machine (created using QEMU/libvirt, LXD, Multipass, etc.), it will likely only work on Intel-based processors due to how vPMU (Virtual Performance Monitoring Unit) works.
```

Comment:
```suggestion
* If you plan to follow along, we recommend setting up a test environment for this experiment.  The ideal setup involves using a bare metal machine because it's the more direct way to collect the performance metrics.  If you would like to use a virtual machine (created using QEMU/libvirt, LXD, Multipass, etc.), it will likely only work on Intel-based processors due to how Virtual Performance Monitoring Unit (vPMU) works.
```

Reply:
It's becoming preferred to put the acronym in brackets after the full name of whatever the thing is - putting the acronym first implies that the reader ought to already know what the acronym stands for